clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -581,11 +597,12 @@ CVarSet
   return Persist;
 }
 
+
 void ConstraintResolver::storePersistentConstraints(clang::Expr *E,
                                                     CVarSet &Vars) {
-  auto PSL = PersistentSourceLoc::mkPSL(E, *Context);
   // Store only if the PSL is valid.
-  if (PSL.valid()) {
+  auto PSL = PersistentSourceLoc::mkPSL(E, *Context);
+  if (PSL.valid() && Rewriter::isRewritable(E->getBeginLoc())){
```

Comment:
I am wondering about the `Rewriter::isRewritable(E->getBeginLoc()))` you are adding here. I think you are adding it because you are concerned that the cast could be in a macro or something like that. But: Why is this check *not* needed for other PVConstraints we might later want to rewrite? Or maybe it is present, e.g., in `addVariable` or something like that?

Reply:
The motivation for this change was an error in vsftpd caused by a macro in the standard library (`isdigit`).  This is a reduced snippet demonstrating  the problem.

```c
extern int **foo(void);
# define bar(c) (*foo())[(int) c]
void test(int a) {
  bar(a);
}
```

This caused an assert to fail during rewriting because the cast expression `(int) c`, had two constraint variables, when it should have had 0 or 1 variable (0 in this case, because it's not a pointer type). This happened because every expression inside the macro has the same source location, so the variables generated for `*foo()` were being treated as if they were constraints for the cast.

I looked into to this, and it looks to me like this check is actually something that we should have had in place when first adding the persistent constraints storage. This example causes a compilation error because `(0,x)` and `(0,y)` have the same source location even though one should be checked while the other is unchecked. 

```c
int *x = (int*) 0;
int *y = (int*) 1;
# define baz(e,f) (e = (0,x));(f = (0,y))

void test(int *c, int *d) {
  baz(c,d);
}
```
Conversion without  check:
```c
_Ptr<int> x =  (_Ptr<int> ) 0;
int *y = (int*) 1;
# define baz(e,f) (e = (0,x));(f = (0,y))

void test(_Ptr<int> c, _Ptr<int> d) {
  baz(c,d);
}
```

Because the expression `(0,y)` has the same source location as `(0,x)`, `hasPeristentConstraintVars` will see that there already exists a (incorrect) constraint variable for `(0,y)` . This variable is returned so, `d` ends up with the same constraints as `c` and they are both converted to `_Ptr`.

---

As for why this isn't  done for variables, any variable declaration that appears in a macro is unconditionally constrained to wild in `addVariable`. This isn't done for expressions in the persistent constraint map. Variable rewriting also doesn't make any assumptions about the size of the constraint variable sets, so if a related issue exists, it would not show up in assert fails.

Reply:
Ok makes sense. Thanks for the explanation. If you can think of a way to get these points into the code comments, that would be great. :)